### ISSUE TITLE
[Infrastructure Request] Set up S3 bucket and service account for DHP team

### ISSUE THREAD

**Issue by yonidoronpeters** [department-of-veterans-affairs/va.gov-team#41598]
## Description
The DHP team needs an encrypted S3 bucket (for dev, staging, and prod) and 1 or 2 service accounts with access to the buckets.

## Background/context 
The @department-of-veterans-affairs/digital-health-platform team needs access to an S3 bucket. The bucket will store Veteran bearer and refresh tokens, and therefore needs to be encrypted with a key from KMS accessible only to our team + a service account which will write new tokens to said bucket.
The S3 bucket will also need to be accessed by another AWS service account originating from [VA Lighthouse](https://developer.va.gov/) with read permissions and access to the key in KMS.

---
## Tasks
For each environment (dev, staging, prod):
- [x] Create S3 bucket 
- [x] Create KMS key for the DHP team
- [x] Create service account with write permissions to S3 bucket and get access to KMS key
- [x] Create service account with read permissions to S3 bucket and get access to KMS Key (this service account will be used by an application in VA Lighthouse)

## Acceptance Criteria 
- [x] All tasks above are complete

---
### Reminders
- [x] Please attach your team label and any other appropriate label(s) (operations, devops, and needs-grooming will automatically be applied as part of the template)
- [x] Please connect to an epic (this will typically be done by the Platform Operations PM or TL)

Please comment with any questions/concerns/feedback. Thanks!

**Comment by oseasmoran73** [comment]
Hi @yonidoronpeters ,

Who is your devops resource contact on your team? We are trying to move towards enabling your devops point of contact.

Also, for future reference, tickets requiring devops/infrastructure service usually scoped/triage by @jhouse-solvd and not directly assigned to one of our members

**Comment by jhouse-solvd** [comment]
@yonidoronpeters - That's helpful info. We'll propose this for the next sprint and scope appropriately given your current need. 

If a follow-up issue is needed, we'll let you know so that you can put in the request.

**Comment by jhouse-solvd** [comment]
Planning to discuss w/ the Platform Infrastructure team on Friday.

**Comment by yonidoronpeters** [comment]
> Planning to discuss w/ the Platform Infrastructure team on Friday.

Thanks for the update @jhouse-solvd!

**Comment by mchaffee-tw** [comment]
@jhouse-solvd Yoni's out today, but please don't hesitate to reach out to michael.chaffee@va.gov or aurelio.nogueira@va.gov if we can answer any questions in real time.  Thanks!

**Comment by jhouse-solvd** [comment]
Hello all. I plan to follow up tomorrow with ETA.

**Comment by yonidoronpeters** [comment]
Hi @jhouse-solvd, just wanted to check in if you have an update on the ETA. We're on the edge of our seat 😄

**Comment by yonidoronpeters** [comment]
@jhouse-solvd Hello! just checking in to see if you have any updates on this request? we are getting to a point where this is blocking other work we have planned. Please let me or @mchaffee-tw know if we can help move this request forward in any way. Thank you 🙏

**Comment by jhouse-solvd** [comment]
@yonidoronpeters - Forgive me for the delayed response. I'm chatting with the team today to see if we can move this forward and unblock your team. We appreciate your continued patience and want to get this done for you.

**Comment by jsgarmon** [comment]
> The S3 bucket will also need to be accessed by another AWS service account originating from VA Lighthouse with read permissions and access to the key in KMS.

@yonidoronpeters : please provide the full ARN for this service account (I'm guessing this is in a separate AWS account)

**Comment by jsgarmon** [comment]
Some additional questions:
- will the new service account be used in GitHub Actions, or used in some other process?
- is the bucket name `dhp-<env>-bearer-tokens` descriptive enough? Do you have a preference for how the bucket is named?

**Comment by mchaffee-tw** [comment]
@jsgarmon Regarding the ARN for the service account...  Lighthouse DI team wants the ARN for the bucket before they create the service account.  In order to resolve the circular dependency, might it be possible to provision the bucket with access only from VFS, and then work out the Lighthouse-side access later?

If it's ever simpler to talk this through on a call, we can be super-available - thanks so much for your help 🙂

**Comment by jsgarmon** [comment]
I can go ahead and create the bucket, but the ARN for the service account needs to be in the bucket policy for the account to access the bucket. The converse isn't true, i.e., the account does not need to be "aware" of the bucket for permission purposes.

I'll put in the PR in a bit, should be created later today.

**Comment by jsgarmon** [comment]
This is complete. The PR above created 3 buckets:
```
dsva-vagov-dev-dhp-bearer-tokens
dsva-vagov-staging-dhp-bearer-tokens
dsva-vagov-prod-dhp-bearer-tokens
```
as well as an uploader service account `vagov-dhp-bucket-uploader` and a KMS key with alias `dsva-vagov-dhp`.

Once I have the ARN for the external service account, I can add that to the permissions for the KMS key and S3 buckets.

**Comment by yonidoronpeters** [comment]
🙌  That's great news - Thank you @jsgarmon! Do you have the ARN for the S3 bucket?

**Comment by jsgarmon** [comment]
arn:aws-us-gov:s3:::dsva-vagov-dev-dhp-bearer-tokens
arn:aws-us-gov:s3:::dsva-vagov-staging-dhp-bearer-tokens
arn:aws-us-gov:s3:::dsva-vagov-prod-dhp-bearer-tokens

**Comment by jsgarmon** [comment]
API keys for the new service account can be found in:
```
/dsva-vagov/vets-api/dhp/vagov-dhp-bucket-uploader/AWS_ACCESS_KEY_ID
/dsva-vagov/vets-api/dhp/vagov-dhp-bucket-uploader/AWS_SECRET_ACCESS_KEY
```

**Comment by jhouse-solvd** [comment]
@yonidoronpeters - Can you confirm that this is working on your side? If so... 

_@jsgarmon - Can you check off tasks and AC and close this out once Yoni confirms?_

**Comment by yonidoronpeters** [comment]
@jhouse-solvd we are trying to confirm this is working as expected. Also waiting on the service accounts from the LDHI side that need access to these buckets. Then the policy will have to change again.

@jsgarmon we realized that the service account on LHDI will need both read & write access to the S3 buckets. Would you like me to change the issue description above to reflect that fact?

**Comment by jsgarmon** [comment]
@yonidoronpeters I'll simply add the external account to the list of principals, which will give it full read write perms. No need to update the ticket.

**Comment by yonidoronpeters** [comment]
@jsgarmon awesome - thanks for that context! 
So, we realized that in the sake of better security practices, we actually need to have better separation of different envs in terms of access to the S3 buckets. Would it be possible to create 2 more service accounts and 2 more KMS keys, and separate the access by environment? 
For example: create another SA with access to the dev s3 bucket only. And create another KMS key which encrypts the dev s3. Then do the same for staging and prod (we can reuse the already created SA and key for one env and limit its access to that single environment).
If this is too much additional work for this github issue, we can create a separate issue with this request. Let us know!

**Comment by yonidoronpeters** [comment]
We also confirmed that we can write and read to the s3 bucket with the supplied SA.

**Comment by yonidoronpeters** [comment]
That's great, thank you so much @jsgarmon!

The roles that need read+write access to the S3 buckets from the Lighthouse side have been created
```
Needs dev access: arn:aws-us-gov:iam::461893774470:role/project/project-ldx/ldx-nonprod-1-digital-health-platform-customer-role-dev
Needs staging access: arn:aws-us-gov:iam::461893774470:role/project/project-ldx/ldx-nonprod-1-digital-health-platform-customer-role-qa-sandbox
Needs prod access: arn:aws-us-gov:iam::461889958881:role/project/project-ldx/ldx-prod-1-digital-health-platform-customer-role
```

**Comment by jsgarmon** [comment]
This has been completed. 3 environment-specific service accounts have been created:
```
arn:aws-us-gov:iam::008577686731:user/service_account/vagov-dev-dhp-bucket-uploader
arn:aws-us-gov:iam::008577686731:user/service_account/vagov-staging-dhp-bucket-uploader
arn:aws-us-gov:iam::008577686731:user/service_account/vagov-prod-dhp-bucket-uploader
```
along with 3 environment specific KMS keys. Bucket and KMS policies have been updated to allow these service accounts and the roles from the Lighthouse account.

Please test and let me know if more needs to be done on this.

**Comment by yonidoronpeters** [comment]
Thank you so much @jsgarmon! we'll test and update shortly

**Comment by oseasmoran73** [comment]
Looks like acceptance criteria has been met. 

@yonidoronpeters Have you had to test out the changes John has made? If they are met, can we close this ticket out?

**Comment by yonidoronpeters** [comment]
@oseasmoran73 We talked about this at standup today and will try to verify today or tomorrow. 

@jsgarmon are the api keys for all service accounts the same?

**Comment by yonidoronpeters** [comment]
@jhouse-solvd confirmed the AC from our end - thanks again team!
